editor/src/core/shared/deep-equality-instances.ts

```diff
@@ -0,0 +1,188 @@
+    newTransientState.fileState,
+  )
+  const fileState = fileStateResult.value
+  areEqual = areEqual && fileStateResult.areEqual
+
+  if (areEqual) {
+    return keepDeepEqualityResult(oldTransientState, true)
+  } else {
+    const transientState: TransientCanvasState = transientCanvasState(
+      selectedViews,
+      highlightedViews,
+      fileState,
+    )
+    return keepDeepEqualityResult(transientState, false)
+  }
+}
+
+export function DerivedStateKeepDeepEquality(
+  oldDerivedState: DerivedState,
+  newDerivedState: DerivedState,
+): KeepDeepEqualityResult<DerivedState> {
+  let areEqual: boolean = true
+
+  const navigatorTargetsResult = TemplatePathArrayKeepDeepEquality(
+    oldDerivedState.navigatorTargets,
+    newDerivedState.navigatorTargets,
+  )
+  const navigatorTargets = navigatorTargetsResult.value
+  areEqual = areEqual && navigatorTargetsResult.areEqual
+
+  const visibleNavigatorTargetsResult = TemplatePathArrayKeepDeepEquality(
+    oldDerivedState.visibleNavigatorTargets,
+    newDerivedState.visibleNavigatorTargets,
+  )
+  const visibleNavigatorTargets = visibleNavigatorTargetsResult.value
+  areEqual = areEqual && visibleNavigatorTargetsResult.areEqual
+
+  let canvasAreEqual: boolean = true
+  const descendantsOfHiddenInstancesResult = TemplatePathArrayKeepDeepEquality(
+    oldDerivedState.canvas.descendantsOfHiddenInstances,
+    newDerivedState.canvas.descendantsOfHiddenInstances,
+  )
+  const descendantsOfHiddenInstances = descendantsOfHiddenInstancesResult.value
+  canvasAreEqual = canvasAreEqual && descendantsOfHiddenInstancesResult.areEqual
+
+  const controlsResult = HigherOrderControlArrayKeepDeepEquality(
+    oldDerivedState.canvas.controls,
+    newDerivedState.canvas.controls,
+  )
+  const controls = controlsResult.value
+  canvasAreEqual = canvasAreEqual && controlsResult.areEqual
+
+  const transientStateResult = TransientCanvasStateKeepDeepEquality(
+    oldDerivedState.canvas.transientState,
+    newDerivedState.canvas.transientState,
+  )
+  const transientState = transientStateResult.value
+  canvasAreEqual = canvasAreEqual && transientStateResult.areEqual
+
+  let canvas: DerivedState['canvas']
+  if (canvasAreEqual) {
+    canvas = oldDerivedState.canvas
+  } else {
+    canvas = {
+      descendantsOfHiddenInstances: descendantsOfHiddenInstances,
+      controls: controls,
+      transientState: transientState,
+    }
+  }
+  areEqual = areEqual && canvasAreEqual
+
+  const elementWarningsEquality: KeepDeepEqualityCall<
+    DerivedState['elementWarnings']
+  > = createCallFromIntrospectiveKeepDeep()
+  const elementWarningsResult = elementWarningsEquality(
+    oldDerivedState.elementWarnings,
+    newDerivedState.elementWarnings,
+  )
+  const elementWarnings = elementWarningsResult.value
+  areEqual = areEqual && elementWarningsResult.areEqual
+
+  if (areEqual) {
+    return keepDeepEqualityResult(oldDerivedState, true)
+  } else {
+    const derivedState: DerivedState = {
```

Comment:
this is probably gross over engineering, but if the derived state changes, whoever will touch this might not see that they need to mutate `areEqual`. if  we'd have a combiner that takes `KeepDeepEqualityResult`s and map them into an object, it could automatically read the `KeepDeepEqualityResult.areEqual`s

Reply:
Ideally I'd implement an applicative-like set of functions as a I did for `Either`. This is just a first swing at the problem.

Reply:
Hmmm I agree here. It might make sense in this case to move `DerivedStateKeepDeepEquality` next to the `deriveState` function to clear that up (and possibly likewise with the other specific deep equality functions)